src/lrf.rs

```diff
@@ -68,173 +68,217 @@ pub enum RestorationFilter {
 
 impl RestorationFilter {
   pub fn default() -> RestorationFilter {
-    RestorationFilter::None {}
+    RestorationFilter::None
   }
 }
 
-fn sgrproj_box_filter(f: &mut[[i32; 1536]; 64], r: isize, eps: isize, clipped_cfg: &PlaneConfig,
-                      stripe_y: isize, stripe_h: isize, y: isize, h: isize,
-                      x: usize, w: usize,
-                      cdeffed: &Plane, deblocked: &Plane,
-                      bit_depth: usize, pass: usize) {
+fn sgrproj_box_ab(af: &mut[i32; 64+2],
+                  bf: &mut[i32; 64+2],
+                  r: isize, eps: isize,
+                  clipped_cfg: &PlaneConfig,
+                  x: isize,
+                  stripe_y: isize, stripe_h: isize, y: isize, h: isize,
+                  cdeffed: &Plane, deblocked: &Plane,
+                  bit_depth: usize) {
+
+
   let n = ((2*r + 1) * (2*r + 1)) as i32;
+  let one_over_n = ((1 << SGRPROJ_RECIP_BITS) + n/2 ) / n;
   let n2e = n*n*eps as i32;
   let s = ((1 << SGRPROJ_MTABLE_BITS) + n2e/2) / n2e;
-  assert!(w <= 1536);
-  assert!(h <= 64);
-  let mut af:[[i32; 1536+2]; 64+2] = [[0; 1536+2]; 64+2];
-  let mut bf:[[i32; 1536+2]; 64+2] = [[0; 1536+2]; 64+2];
-  for i in -1..1+h {
-    for j in -1..1+w as isize{
-      let mut a:i32 = 0;
-      let mut b:i32 = 0;
-
-      let xn = cmp::min(r+1, clipped_cfg.width as isize - x as isize - j);
-      for yi in y+i-r..=y+i+r {
-        let mut src_plane: &Plane;
-        let ly;
-        if yi < stripe_y {
-          ly = cmp::max(clamp(yi, 0, clipped_cfg.height as isize - 1), y - 2) as usize;
-          src_plane = deblocked;
-        } else if yi < stripe_y + stripe_h {
-          ly = clamp(yi, 0, clipped_cfg.height as isize - 1) as usize;
-          src_plane = cdeffed;
-        } else {
-          ly = cmp::min(clamp(yi, 0, clipped_cfg.height as isize - 1), y+h+1) as usize;
-          src_plane = deblocked;
-        }
+  let xn = cmp::min(r+1, clipped_cfg.width as isize - x);
+  for row in -1..1+h {
+    let mut a:i32 = 0;
+    let mut b:i32 = 0;
 
-        for _xi in -r..-(x as isize)-j {
-          let c = src_plane.p(0, ly) as i32;
-          a += c*c;
-          b += c;
-        }
-        for xi in cmp::max(-r, -(x as isize)-j)..xn {
-          let c = src_plane.p((xi + x as isize + j) as usize, ly) as i32;
-          a += c*c;
-          b += c;
-        }
-        for _xi in xn..r+1 {
-          let c = src_plane.p(clipped_cfg.width - 1, ly) as i32;
-          a += c*c;
-          b += c;
-        }
-      }
-      a = a + (1 << 2*(bit_depth-8) >> 1) >> 2*(bit_depth-8);
-      let d = b + (1 << bit_depth - 8 >> 1) >> bit_depth - 8;
-      let p = cmp::max(0, a*(n as i32) - d*d);
-      let z = p*s + (1 << SGRPROJ_MTABLE_BITS >> 1) >> SGRPROJ_MTABLE_BITS;
-      let a2 = if z >= 255 {
-        256
-      } else if z == 0 {
-        1
+    for yi in y+row-r..=y+row+r {
+      let mut src_plane: &Plane;
+      let ly;
+      if yi < stripe_y {
```

Comment:
`src_plane` can stay immutable or I'm misreading?

You may do 
``` rust
let (ly, src_plane) = if yi < stripe_y {
  (cmp::max(clamp(yi, 0, clipped_cfg.height as isize - 1), y - 2) as usize,
  deblocked)
} ...
```

and such but probably looks simpler with the assignment.